sycl/tools/sycl-trace/collector.cpp

```diff
@@ -8,21 +8,146 @@
+  auto LastSlashPos = Path.find_last_of('/');
+
+  return Path.substr(0, LastSlashPos);
+}
+
+class CollectorLibraryWrapper {
+  typedef void (*InitFuncType)();
+  typedef void (*FinishFuncType)();
+  typedef void (*CallbackFuncType)(uint16_t, xpti::trace_event_data_t *,
+                                   xpti::trace_event_data_t *, uint64_t,
+                                   const void *);
+  typedef void (*SetIndentLvlFuncType)(int);
+
+public:
+  CollectorLibraryWrapper(const std::string &LibraryName)
+      : MLibraryName(LibraryName){};
+  ~CollectorLibraryWrapper() { clear(); };
+
+  const std::string InitFuncName = "init";
+  const std::string FinishFuncName = "finish";
+  const std::string CallbackFuncName = "callback";
+  const std::string IndentFuncName = "setIndentationLevel";
+
+  bool initPrinters() {
+    const std::string CollectorDir = getCurrentDSODir() + "/" + MLibraryName;
+    MHandle = dlopen(CollectorDir.c_str(), RTLD_LAZY);
+    if (!MHandle) {
+      std::cerr << "Cannot load library: " << dlerror() << '\n';
+      return false;
+    }
+    auto ExportSymbol = [&](void *&FuncPtr, const std::string &FuncName) {
+      FuncPtr = dlsym(MHandle, FuncName.c_str());
+      if (!FuncPtr) {
+        std::cerr << "Cannot export symbol: " << dlerror() << '\n';
+        return false;
+      }
+      return true;
+    };
+    if (!ExportSymbol(MInitPtr, InitFuncName) ||
+        !ExportSymbol(MFinishPtr, FinishFuncName) ||
+        !ExportSymbol(MSetIndentationLevelPtr, IndentFuncName) ||
+        !ExportSymbol(MCallbackPtr, CallbackFuncName)) {
+      clear();
+      return false;
+    }
+
+    if (MIndentationLevel)
+      ((SetIndentLvlFuncType)MSetIndentationLevelPtr)(MIndentationLevel);
+
+    ((InitFuncType)MInitPtr)();
+
+    return true;
+  }
+
+  void finishPrinters() {
+    if (MHandle)
+      ((FinishFuncType)MFinishPtr)();
+  }
+
+  void setIndentationLevel(int Level) {
+    MIndentationLevel = Level;
+    if (MHandle)
+      ((SetIndentLvlFuncType)MSetIndentationLevelPtr)(MIndentationLevel);
+  }
+
+  void callback(uint16_t TraceType, xpti::trace_event_data_t *Parent,
+                xpti::trace_event_data_t *Event, uint64_t Instance,
+                const void *UserData) {
+    // Not expected to be called when MHandle == NULL since we should not be
+    // subscribed if init failed. Although still do the check for sure.
+    if (MHandle)
+      ((CallbackFuncType)MCallbackPtr)(TraceType, Parent, Event, Instance,
+                                       UserData);
+  }
+
+  void clear() {
+    MInitPtr = NULL;
+    MFinishPtr = NULL;
+    MCallbackPtr = NULL;
+    MSetIndentationLevelPtr = NULL;
+
+    if (MHandle)
+      dlclose(MHandle);
+    MHandle = NULL;
+  }
+
+private:
+  std::string MLibraryName;
+  int MIndentationLevel = 0;
+
+  void *MHandle = NULL;
+
+  void *MInitPtr = NULL;
+  void *MFinishPtr = NULL;
+  void *MCallbackPtr = NULL;
+  void *MSetIndentationLevelPtr = NULL;
```

Comment:
nit: should we use `nullptr` instead of `NULL`?

Reply:
"If handle does not refer to a valid object opened by [dlopen()](https://pubs.opengroup.org/onlinepubs/009696899/functions/dlopen.html), or if the named symbol cannot be found within any of the objects associated with handle, dlsym() shall return NULL."
I assume there should be no difference if I change it to nullptr. I am ok doing that. Although I tried to work in terms of used API.